lib/rhea-promise/index.ts

```diff
@@ -0,0 +1,662 @@
+ * @param {Session} session The amqp session object on which the receiver link needs to be established.
+ * @param {ReceiverOptions} [options] Options that can be provided while creating an amqp receiver.
+ * @return {Promise<Receiver>} Promise<Receiver>
+ * - **Resolves** the promise with the Receiver object when rhea emits the "receiver_open" event.
+ * - **Rejects** the promise with an AmqpError when rhea emits the "receiver_close" event while trying
+ * to create an amqp receiver.
+ */
+export async function createReceiver(session: any, options?: ReceiverOptions): Promise<any> {
+  return new Promise((resolve, reject) => {
+    if (!session || (session && typeof session !== "object")) {
+      reject(new Error("session is a required parameter and must be of tyepe 'object'."));
+    }
+
+    const receiver = session.attach_receiver(options);
+
+    function onOpen(context: any): void {
+      receiver.removeListener("receiver_open", onOpen);
+      receiver.removeListener("receiver_close", onClose);
+      process.nextTick(() => {
+        debug(`Resolving the promise with amqp receiver "${receiver.name}".`);
+        resolve(receiver);
+      });
+    }
+
+    function onClose(context: Context): void {
+      receiver.removeListener("receiver_open", onOpen);
+      receiver.removeListener("receiver_close", onClose);
+      debug(`Error occurred while creating a receiver over amqp connection.`, context.receiver.error);
+      reject(context.receiver.error);
+    }
+
+    receiver.once("receiver_open", onOpen);
+    receiver.once("receiver_close", onClose);
+  });
+}
+
+/**
+ * Closes the amqp receiver.
+ * @param {Receiver} receiver The amqp receiver that needs to be closed.
+ * @return {Promise<void>} Promise<void>
+ * - **Resolves** the promise when rhea emits the "receiver_close" event.
+ * - **Rejects** the promise with an AmqpError when rhea emits the
+ * "receiver_error" event while trying to close an amqp receiver.
+ */
+export async function closeReceiver(receiver: any): Promise<void> {
+  return new Promise<void>((resolve, reject) => {
+    if (!receiver || (receiver && typeof receiver !== "object")) {
+      reject(new Error("receiver is a required parameter and must be of tyepe 'object'."));
+    }
+
+    if (receiver.is_open()) {
+      function onClose(context: Context): void {
+        receiver.removeListener("receiver_close", onClose);
+        process.nextTick(() => {
+          debug("Resolving the promise as the amqp receiver has been closed.");
+          resolve();
+        });
+      }
+
+      function onError(context: Context): void {
+        receiver.removeListener("receiver_error", onError);
+        debug(`Error occurred while closing amqp receiver.`, context.receiver.error);
+        reject(context.receiver.error);
+      }
+
+      receiver.once("receiver_close", onClose);
+      receiver.once("receiver_error", onError);
+      receiver.close();
+    } else {
+      resolve();
+    }
+  });
+}
+
+/**
```

Comment:
From here onwards I have added type definitions for options that can be provided while establishing a connection or creating a sender, receiver, session. Please let me know if I am missing anything or have put something incorrectly. I have used the documentation in rhea (Readme.md) and the [AMQP 1.0 pdf](http://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-complete-v1.0-os.pdf) for providing description for some of the properties.